jupyter_server/services/kernels/handlers.py

```diff
@@ -449,55 +457,102 @@ def subscribe(value):
+            if self.selected_subprotocol == "v1.kernel.websocket.jupyter.org":
+                self.session.send_raw(stream, msg_list)
+            else:
+                self.session.send(stream, msg)
+
+    def get_part(self, field, value, msg_list):
+        if value is None:
+            field2idx = {
+                "header": 0,
+                "parent_header": 1,
+                "content": 3,
+            }
+            value = self.session.unpack(msg_list[field2idx[field]])
+        return value
 
     def _on_zmq_reply(self, stream, msg_list):
         idents, fed_msg_list = self.session.feed_identities(msg_list)
-        msg = self.session.deserialize(fed_msg_list)
 
-        parent = msg["parent_header"]
-
-        def write_stderr(error_message):
-            self.log.warning(error_message)
-            msg = self.session.msg(
-                "stream", content={"text": error_message + "\n", "name": "stderr"}, parent=parent
-            )
-            msg["channel"] = "iopub"
-            self.write_message(json.dumps(msg, default=json_default))
+        if self.selected_subprotocol == "v1.kernel.websocket.jupyter.org":
+            msg = {"header": None, "parent_header": None, "content": None}
+        else:
+            msg = self.session.deserialize(fed_msg_list)
 
         channel = getattr(stream, "channel", None)
-        msg_type = msg["header"]["msg_type"]
+        parts = fed_msg_list[1:]
+
+        self._on_error(channel, msg, parts)
 
-        if channel == "iopub" and msg_type == "error":
-            self._on_error(msg)
+        if self._limit_rate(channel, msg, parts):
+            return
+
+        if self.selected_subprotocol == "v1.kernel.websocket.jupyter.org":
+            super(ZMQChannelsHandler, self)._on_zmq_reply(stream, parts)
+        else:
+            super(ZMQChannelsHandler, self)._on_zmq_reply(stream, msg)
+
+    def write_stderr(self, error_message, msg):
+        self.log.warning(error_message)
+        err_msg = self.session.msg(
+            "stream",
+            content={"text": error_message + "\n", "name": "stderr"},
+            parent=msg["parent_header"],
+        )
+        if self.selected_subprotocol == "v1.kernel.websocket.jupyter.org":
+            bin_msg = serialize_msg_to_ws_v1(err_msg, "iopub", self.session.pack)
+            self.write_message(bin_msg, binary=True)
+        else:
+            err_msg["channel"] = "iopub"
+            self.write_message(json.dumps(err_msg, default=json_default))
 
-        if (
-            channel == "iopub"
-            and msg_type == "status"
-            and msg["content"].get("execution_state") == "idle"
-        ):
+    def _limit_rate(self, channel, msg, msg_list):
+        if not (channel == "iopub" and self.limit_rate):
+            return False
+
+        msg["header"] = self.get_part("header", msg["header"], msg_list)
+        msg["content"] = self.get_part("content", msg["content"], msg_list)
```

Comment:
I think you don't need to parse the content unless the header message type is `status`, right? Status messages are relatively lightweight, so if you delay getting the content until after you know it is a status message, you avoid parsing very expensive `content` of other messages.

Reply:
Good catch!